src/modules/live-execution-tracker/live-execution-tracker.ts

```diff
@@ -533,68 +535,74 @@ export class LiveExecutionTracker {
 
-    const elementHasIncomingElements: boolean = incomingElementsAsIModdleElement === undefined;
-    if (elementHasIncomingElements) {
-      return elementWithIncomingElements;
-    }
-
-    for (const incomingElement of incomingElementsAsIModdleElement) {
-      const incomingElementAsShape: IShape = this._elementRegistry.get(incomingElement.id);
-      const sourceOfIncomingElement: IShape = incomingElementAsShape.source;
-
-      const incomignElementHasNoSource: boolean = sourceOfIncomingElement === undefined;
-      if (incomignElementHasNoSource) {
+      const outgoingElementHasNoTarget: boolean = targetOfOutgoingElement === undefined;
+      if (outgoingElementHasNoTarget) {
         continue;
       }
 
-      const previousElementIsTask: boolean = sourceOfIncomingElement.type.includes('Task');
+      const outgoingElementHasNoActiveToken: boolean = !this._hasElementActiveToken(targetOfOutgoingElement.id);
+      const targetOfOutgoingElementHasNoTokenHistory: boolean = !this._hasElementTokenHistory(targetOfOutgoingElement.id, tokenHistoryGroups);
 
-      if (previousElementIsTask) {
-        const elementHasActiveToken: boolean = this._hasElementActiveToken(sourceOfIncomingElement.id);
-        const sourceOfIncomingElementHasNoTokenHistory: boolean = !this._hasElementTokenHistory(sourceOfIncomingElement.id, tokenHistoryGroups);
+      if (outgoingElementHasNoActiveToken && targetOfOutgoingElementHasNoTokenHistory) {
+        continue;
+      }
 
-        if (elementHasActiveToken || sourceOfIncomingElementHasNoTokenHistory) {
-          continue;
-        }
+      const outgoingElementIsSequenceFlow: boolean = outgoingElementAsShape.type === 'bpmn:SequenceFlow';
+      if (outgoingElementIsSequenceFlow) {
+        const tokenHistoryForTarget: TokenHistoryEntry = tokenHistoryGroups[targetOfOutgoingElement.id][0];
+        const previousFlowNodeInstanceIdOfTarget: string = tokenHistoryForTarget.previousFlowNodeInstanceId;
 
-        elementWithIncomingElements.push(incomingElementAsShape);
+        const tokenHistoryForElement: TokenHistoryEntry = tokenHistoryGroups[element.id][0];
+        const flowNodeInstanceIdOfElement: string = tokenHistoryForElement.flowNodeInstanceId;
 
-      } else {
-        elementWithIncomingElements.push(incomingElementAsShape);
+        // This is needed because the ParallelGateway only knows the flowNodeId of the first element that reaches the ParallelGateway
+        const targetOfOutgoingElementIsGateway: boolean = targetOfOutgoingElement.type === 'bpmn:ParallelGateway';
 
-        const sourceHasNoTokenHistory: boolean = !this._hasElementTokenHistory(sourceOfIncomingElement.id, tokenHistoryGroups);
-        if (sourceHasNoTokenHistory) {
-          elementWithIncomingElements.push(sourceOfIncomingElement);
+        const sequenceFlowWasExecuted: boolean = previousFlowNodeInstanceIdOfTarget === flowNodeInstanceIdOfElement;
+        if (sequenceFlowWasExecuted || targetOfOutgoingElementIsGateway) {
```

Comment:
Theres is another business rule. Extract this please.

```suggestion
       const needToAddToOutgoingElements: boolean  = sequenceFlowWasExecuted || targetOfOutgoingElementIsGateway
        if (needToAddToOutgoingElements) {
```